### ISSUE TITLE
Caching of often used reference/relation data

### ISSUE THREAD

**Issue by marceloverdijk** [prisma/prisma#2109]
## Problem

In terms of application performance often used reference data (like countries, categories, or even authors) could be cached.

E.g. when doing something like:

```
const allUsers = await prisma.user.findMany({
  include: { country: true },
})
```

Now the country is always queried using Prisma. But these countries could live in an application wide Cache as it's quite static.

## Solution

Prisma could implement a cache layer similar as Hibernate (Java) second-level cache (see a brief description of it here: https://www.baeldung.com/hibernate-second-level-cache).

Note implementing such a cache layer is not simple. E.g. think about evicting/updating cache entries when data is updated/deleted.

## Alternatives

Alternatively - without such a caching layer - users have to build such a solution itself.

## Additional context

https://www.baeldung.com/hibernate-second-level-cache
https://hazelcast.com/glossary/hibernate-second-level-cache/

**Comment by Sytten** [comment]
My suggestion on caching would be to provide helpers like redis and in memory, but  let the use be explicit to the user using a `.cache(key)` in the call.

**Comment by marceloverdijk** [comment]
Yes the cache implementation should be abstract (like Hibernate also supports).

**Comment by marceloverdijk** [comment]
Not automatic, but this is useful and a technique I currently use:
https://discord.com/channels/830184174198718474/830184175176122389/889911170583130132

**Comment by TasinIshmam** [comment]
Hi folks 👋, 

Tasin from Prisma here. 

We are launching a [Design Partner Program](https://prismaio.notion.site/Prisma-Design-Partner-Program-a286c25e47fb44e199038ca71dce36a1) where external teams can collaborate with us to shape Prisma's future. In this program, we will collaborate with you on defining solutions to the following problems: 

- **Caching**
- Listening and reacting to database changes in real-time
- Serverless workflow orchestration
- Data projections for denormalization or data integration purposes

**If you’re struggling with caching, especially in a serverless environment, you might be a great fit for this program!**  Your team can take advantage of new Prisma features that you won’t have to build or maintain yourselves. As you invest time in partnering with us, we will also offer you free credits once anything coming out of this program reaches production.

**You can [check out the program in more detail](https://prismaio.notion.site/Prisma-Design-Partner-Program-a286c25e47fb44e199038ca71dce36a1) to see the kinds of activities we’re planning. To apply to this program, [fill out the form](https://docs.google.com/forms/d/e/1FAIpQLSdfadMO7qVOMlOgmeYevpM-olpjku2-3sVzMvVQiHITmZf4dA/viewform)**.

We’re looking forward to hearing from you!

**Comment by janpio** [comment]
Am I correct in understanding that the expectation here would be that Prisma, instead of joining the `country` for each user, or doing a separate query to get the country details for one or multiple users, each time you run `prisma.user.findMany`, Prisma would keep a cache of all the `country` it has seen before, and prefer to use that? And then, if all the requests `country` for a query are already known, it would use those cached values instead to craft the object that the query has to return?

**Comment by marceloverdijk** [comment]
Indeed @janpio that's exactly what this proposal is.
